Add `IMicrosoftIdentityAuthenticationDelegatingHandlerFactory` to allow for overriding handlers

MattKotsenas opened the pull request:
Currently, it's very difficult to customize the behavior of the
`DelegatingHandler` classes that add the auth token, as the
`DownstreamWebApiExtensions` `new()` up the handlers directly.

Ideally I would register the handlers in the `IServiceCollection`
directly, and then resolve it from the container, however in the case of
multiple downstream APIs we need to resolve the correct named
configuration from the `IOptionsMonitor`, and I'm not aware of a way to
do that with just the dependency injection abstraction APIs.

As a result, I've instead introduced a public factory interface that
takes the configuration name, which then
allows for creating the handler from the factory class. Users that want
to customize the handlers can then register their own
`IMicrosoftIdentityAuthenticationDelegatingHandlerFactory` 
that creates their own handlers.

jennyf19 commented:
@qetza Do you mind reviewing this as well? thanks.

qetza commented on src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationDelegatingHandlerFactory.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <returns>
+        /// The <see cref="DelegatingHandler"/>.
+        /// </returns>
+        DelegatingHandler CreateAppHandler(IServiceProvider serviceProvider, string? serviceName);
```
The `IServiceProvider` should not be given as a parameter of the method but managed by the class implementing the interface (maybe some implementation doesn't need an IServiceProvider)

MattKotsenas replied:
Good idea! I removed `IServiceProvider` from the interface's signature and added it for constructor injection in the default implementation.

qetza commented on src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationDelegatingHandlerFactory.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <returns>
+        /// The <see cref="DelegatingHandler"/>.
+        /// </returns>
+        DelegatingHandler CreateUserHandler(IServiceProvider serviceProvider, string? serviceName);
```
Same as above 😄

qetza commented on src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationDelegatingHandlerFactory.cs:
```diff
@@ -0,0 +1,73 @@
+    /// <summary>
+    /// The default implementation of <see cref="IMicrosoftIdentityAuthenticationDelegatingHandlerFactory"/>.
+    /// </summary>
+    public class MicrosoftIdentityAuthenticationDelegatingHandlerFactory : IMicrosoftIdentityAuthenticationDelegatingHandlerFactory
```
As this is the default implementation and will be managed by the extension methods the class could be `internal` and named `DefaultMicrosoftIdentityAuthenticationDelegatingHandlerFactory`

qetza commented on src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationDelegatingHandlerFactory.cs:
```diff
@@ -0,0 +1,73 @@
+    /// <summary>
+    /// The default implementation of <see cref="IMicrosoftIdentityAuthenticationDelegatingHandlerFactory"/>.
+    /// </summary>
+    public class MicrosoftIdentityAuthenticationDelegatingHandlerFactory : IMicrosoftIdentityAuthenticationDelegatingHandlerFactory
```
The class should get the `IServiceProvider` as a dependency injection in the constructor and use it internally.

MattKotsenas replied:
👍

qetza commented on src/Microsoft.Identity.Web/DownstreamWebApiSupport/MicrosoftIdentityAuthenticationMessageHandlerHttpClientBuilderExtensions.cs:
```diff
@@ -32,13 +31,12 @@ public static class MicrosoftIdentityAuthenticationMessageHandlerHttpClientBuild
             }
 
             builder.Services.Configure<MicrosoftIdentityAuthenticationMessageHandlerOptions>(serviceName, configuration);
+            builder.Services.AddSingleton<IMicrosoftIdentityAuthenticationDelegatingHandlerFactory, MicrosoftIdentityAuthenticationDelegatingHandlerFactory>();
```
You should use `TryAddSingleton` on `IMicrosoftIdentityAuthenticationDelegatingHandlerFactory` so that a user can configure his own implementation before and we will not override it.

As this is done in all extension methods maybe put this in a `AddMicrosoftIdentityAuthenticationHandlerCore` private method called by all other extension methods so that common configuration is shared.

qetza reviewed (changes requested):
Hi @MattKotsenas,
This is a good idea, i've added some comment.

MattKotsenas commented:
Thanks for all the feedback @qetza! I believe I've addressed all your comments. For some reason clicking the "re-request review" button doesn't seem to be doing anything for me, so I'm tagging you here to take another look. Thanks again!

jmprieur commented on src/Microsoft.Identity.Web/DownstreamWebApiSupport/IMicrosoftIdentityAuthenticationDelegatingHandlerFactory.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    /// <summary>
+    /// Interface to a class that provides the <see cref="DelegatingHandler"/> that
+    /// addes an authorization header with a token for the application.
```
```suggestion
    /// adds an authorization header with a token for the application.
```

MattKotsenas replied:
Fixed!

jmprieur reviewed (approved):
LGTM

jennyf19 commented:
#1667
